### ISSUE TITLE
Support for issuance

### ISSUE THREAD

**Issue by samuelgoto** [w3c-fedid/digital-credentials#167]
We deliberately left issuance aside while we were trying to figure out presentation, so kicking this issue off just to make sure it doesn't fall through the cracks :)

As thing start to settle with presentation, we should starting thinking how to go about issuance. There are a lot of things and a lot of considerations, so we'll likely want to break this into smaller steps, but kicking this off just to get the ball rolling.

**Comment by OR13** [comment]
In presentations, the verifier wants to select claims for the holder to disclose and be protected against replays with nonces.

In issuance, the issuer wants to be assured the holder controls a key, and has consented to the claims that will be bound to that key... The holder wants to know which media types are supported, and what credential types are supported.

So there is some symmetry around proof of possession, media types, and credential schemas / types (known structures).

There is also: https://openid.github.io/OpenID4VCI/openid-4-verifiable-credential-issuance-wg-draft.html

Are you asking what are the essential parameters a browser API would need to collect in order to submit an issuance request to some backend?

What is the threat model you have in mind for issuance?

Is it necessary for the holder to encrypt their pop key or other credential information to the issuer?

Should the browser or OS be able to observe specific claims or credential types, or should they just see an encrypted request for issuance?

**Comment by samuelgoto** [comment]
Ok, here is something to get the discussion started. WDYT?

```javascript
// Prompts the user to select a wallet to have the digital credential issued
await navigator.credentials.store({
  digital: {
    protocol: "openid4vci",
    data: {
      // the openid4vci request body, omitted for brevity
    },
  }
});
```

**Comment by timcappalli** [comment]
@samuelgoto what's the reasoning for `.store`? The ultimate result is the creation of a digital credential in an external container. I think it should be `.create`, similar to WebAuthn.

**Comment by samuelgoto** [comment]
> @samuelgoto what's the reasoning for .store? The ultimate result is the creation of a digital credential in an external container. I think it should be .create, similar to WebAuthn.

Are you referring to this `TODO` that I left [1]? The code snippet I proposed uses `create()` rather than `store()`, right?

[1]
```javascript
// TODO(goto): should we use the navigator.credentials.store(new DigitalCredential()) instead of create()?
```

**Comment by timcappalli** [comment]
Sorry, looks like I saw an earlier version before it was edited.

**Comment by samuelgoto** [comment]
Question: is there an equivalent of a "un-issuance"? Can a wallet `create` and then invalidate / delete it?

**Comment by samuelgoto** [comment]
// TODO(goto): ask @marcoscaceres what's the difference between the credential manager's `store()` and `create()` call ... 
// it seems like `PublicKeyCredential` uses `create()`, but `PasswordCredential` and `FederatedCredential` use `store()`
// instead ([example](https://developer.mozilla.org/en-US/docs/Web/API/PasswordCredential#examples)).

**Comment by timcappalli** [comment]
> Question: is there an equivalent of a "un-issuance"? Can a wallet create and then invalidate / delete it?

Not typically via the front channel. WebAuthn will soon have some new methods that allow signaling to the client that credentials are no longer valid, but it's a background / opportunistic thing.

> @marcoscaceres what's the difference between the credential manager's store() and create() call?
> it seems like PublicKeyCredential uses create(), but PasswordCredential and FederatedCredential use store() instead > ([example](https://developer.mozilla.org/en-US/docs/Web/API/PasswordCredential#examples)).
>
> Which one should we use here for issuance?

Architecturally, digital credentials and wallets are the same as public key credentials and authenticators. Store is used with user agent internal credential stores. Like authenticators, digital wallets are external, and should use `create`.

**Comment by samuelgoto** [comment]
I've been trying to inform myself on how issuance works, and it seems like there is a pretty tight connection between Issuers and Holders ... meaning, it seems some (enough? all?) Issuers have an out-of-band agreement with Holders and only allow a certain set of enumerated and approved holders ...

That is, am I reading this correctly in that a "wallet chooser" isn't quite correct? -- or only to the extent that the issuer can pick the *specific* wallets that are offered to the user.

**Comment by samuelgoto** [comment]
>  That particular subsets of the ecosystem are more closed does not mean that others aren't more open (they are).

Yeah, that makes sense. I guess, maybe the point that I was trying to make here is that the closed wallet choice is necessary (rather than, as you said, sufficient), and I think we don't yet know how to address that necessity.

**Comment by dlongley** [comment]
> I guess, maybe the point that I was trying to make here is that the closed wallet choice is necessary (rather than, as you said, sufficient), and I think we don't yet know how to address that necessity.

One way this sort of thing can be done with the VC API is for the issuer to ask for a VC from the wallet (where that VC is about the wallet itself) during an exchange, prior to (gating) issuance. Perhaps something similar can be modeled or "announced" in an issuance call via the browser API -- whereby a request is included in the issuance parameters (somewhere in `store()`) that could be evaluated in a similar way that requests in `get()` are / may be to determine applicability.

**Comment by leecam** [comment]
As per discussion on last week's call, some more background on Issuance that could potentially be included in the explainer.

One of the primary motivations for the Digital Credentials API for presentation was the issue of the use of custom schemes by protocols such as OpenID4VP and the references to those protocols in regulations (such as eIDAS2.0). A significant amount of work has gone into updating specifications such as OpenIdD4VP and HAIP to reference the DC API for presentation. Recently we've also seen governments start to make formal references to the DC API for presentation.

While we have made substantial progress with presentation, credential Issuance suffers from the same issues with regard to custom schemes. Protocols such as [OpenID4VCI](https://openid.github.io/OpenID4VCI/openid-4-verifiable-credential-issuance-wg-draft.html) and the HAIP specification still have no choice but to mandate custom schemes to initiate issuance. As such we're seeing this being captured in regulations as these standards move to the final stages.

To address this the DC API should provide a means to allow an issuer website to initiate the issuance of credentials to the user's wallet of choosing. As with presentation, this would unblock Issuance protocols such as OpenID4VCI and specifications such as HAIP to formally reference the DC API as an alternate to custom schemes.

Its proposed that Issuance follow the same API design pattern as presentation, with the same key design principle around protocol agility and extensibility. The API should build upon the Credential Management API. As such we propose the use of ```navigator.credentials.create``` with a new DigitalCredentialCreationOptions extension to [CredentialCreationOptions](https://www.w3.org/TR/credential-management-1/#credentialcreationoptions-dictionary).

Mirroring the presentation call the API could take the following form:

```javascript
const result = await navigator.credentials.create({
  digital: {
    requests: [{
      protocol: "openid4vci",
      data: {
        // the openid4vci request body, omitted for brevity
      }
   }, {
      protocol: "some-other-issuance-protocol",
      data: {
        // the request body, omitted for brevity
      }   
   }],
  }
});
```

The proposal to allow a list of credential offers may require some discussion. The intention is that a given wallet will only offer to process a single request within the list. As such, this implies that each request represents the same credential offer, just offered using different protocols. This is subtly different to the get() call where wallets may offer all the unique credentials that match across all the requests in the list.

The motivation here is to allow issuers to make a single call to make a credential offer, without knowing the protocols supported by the available wallets. For example lets say in the future a new OpenID4VCI v2.0 protocol is released and an Issuer wishes to upgrade to using that protocol, while still supporting legacy wallets. An Issuer can now make a call to offer a credential, such as a driving license, using both OpenID4VCI v1.0 and OpenID4VCI v2.0.

As with the presentation call, a DigitalCredential is returned containing the protocol that was used by the wallet to process the request. Likewise, the returned data object is defined by the protocol. It is expected that many issuance protocols will use this response object to merely convey the success or failure of the issuance request.

**Comment by samuelgoto** [comment]
Just a small update here: we are going to send a spec PR to get this proposed into the spec.

**Comment by marcoscaceres** [comment]
> @marcoscaceres what's the difference between the credential manager's store() and create() call?
> it seems like PublicKeyCredential uses create(), but PasswordCredential and FederatedCredential use store() instead ([example](https://developer.mozilla.org/en-US/docs/Web/API/PasswordCredential#examples)).

Right, it's because there is no way to "create" a `PublicKeyCredential` from, for instance, a HTML form (which `.store()` would support). To create `PublicKeyCredential` you must go through the `.create()` method.

Similarly with a `DigitalCredential`, you can't create it from HTML or anything... you must go through `.create()`.
